File: datastruct/union_find/UnionFind.java

public class UnionFind implements UF {

    private int[] parent; // 存储元素 i 父结点的索引
    private int[] rank; // 以索引 i 为根的树的深度

    public UnionFind(int size){
        parent = new int [size];
        rank = new int[size];
        for(int i = 0; i < size; i++) {
            parent[i] = i;
            rank[i] = 1;
        }
    }

    @Override
    public int getSize() {
        return parent.length;
    }

    // 找到 index 所在集合的根结点
    private int find(int index){
        if(index < 0 || index >= parent.length)
            throw new IndexOutOfBoundsException("Index is out of bound");

        while(index != parent[index]) {
            parent[index] = parent[parent[index]];
            index = parent[index];
        }
        // 另一种路径压缩，通过递归实现将本次查询路径高度压缩为 2
//        if(index != parent[index]) {
//            parent[index] = find(parent[index]);
//        }
        return index;
    }

    // 判断 p，q 是否在一个集合内
    @Override
    public boolean isConnected(int p, int q) {
        return find(p) == find(q);
    }

    // 合并 p，q 所在集合
    @Override
    public void union(int p, int q) {
        int pRoot = find(p);
        int qRoot = find(q);

        if(pRoot == qRoot)
            return;

        if(rank[pRoot] < rank[qRoot]){
            parent[pRoot] = qRoot;
        } else if (rank[qRoot] < rank[pRoot]) {
            parent [qRoot] = pRoot;
        } else {
            parent[qRoot] = pRoot;
            rank[pRoot] += 1;
        }

    }
}

File: datastruct/union_find/UnionFind.py
from UF import UF

class UnionFind(UF):

    def __init__(self, size:int) -> None:
        self.parent = [None] * size
        self.rank = [None] * size
        for i in range(size):
            self.parent[i] = i
            self.rank[i] = 1
    
    def get_size(self) -> int:
        return len(self.parent)
   
    # 找到 index 所在集合的根结点
    def __find(self, index:int) ->int:

        if index < 0 or index >= len(self.parent):
            raise IndexError("Index out of range")

        while index != self.parent[index]:
            self.parent[index] = self.parent[self.parent[index]]
            index = self.parent[index]

        # 第二种路径压缩，通过递归实现将路径高度压缩至2
        # if index != self.parent[index]:
        #     self.parent[index] = self.__find(self.parent[index])
        #     index = self.parent[index]
        return index
    
    # 判断 p，q 是否在一个集合内
    def is_connected(self, p:int, q:int) -> bool:
        return self.parent[p] == self.parent[q]

    # 合并 p，q 所在集合
    def union(self, p:int, q:int) -> None:

        p_root = self.__find(p)
        q_root = self.__find(q)

        if q_root == p_root:
            return
        
        if self.rank[p] < self.rank[q]:
            self.parent[p] = q_root
        elif self.rank[p] > self.rank[q]:
            self.parent[q] = p_root
        else:
            self.parent[q] = p_root
            self.rank[p_root] += 1